File: bin/lib/permissions/matchers.mjs
/**
 * Permission pattern matching functions
 * Handles glob patterns, bash commands, domains, and tool permission strings
 */

import { URL } from 'url';

/**
 * Simple glob pattern matching
 * @param {string} pattern - Glob pattern (supports * and ?)
 * @param {string} str - String to match
 * @returns {boolean} Whether pattern matches string
 */
export function matchGlob(pattern, str) {
  if (typeof pattern !== 'string' || typeof str !== 'string') {
    return false;
  }

  if (pattern.length > 1000) {
    return false;
  }

  try {
    const escaped = pattern.replace(/[.+^${}()|[\]\\]/g, '\\$&');
    const regexPattern = escaped.replace(/\*/g, '.*?').replace(/\?/g, '.');
    const regex = new RegExp(`^${regexPattern}$`);
    return regex.test(str);
  } catch {
    return false;
  }
}

/**
 * Parse tool permission string like "Tool(pattern)"
 * @param {string} toolStr - Tool permission string
 * @returns {Object} Parsed pattern { toolName, ruleContent, type }
 */
export function parseToolPattern(toolStr) {
  const granularMatch = toolStr.match(/^([a-z]+)\((.+)\)$/i);
  if (granularMatch) {
    const toolName = granularMatch[1].toLowerCase();
    const ruleContent = granularMatch[2];

    if (toolName === 'bash') {
      const isWildcard = ruleContent.match(/^([^:]+):\*$/);
      return {
        toolName: 'bash',
        ruleContent: ruleContent.toLowerCase(),
        type: isWildcard ? 'bash_wildcard' : 'bash_exact',
      };
    } else if (['read', 'write', 'edit'].includes(toolName)) {
      return {
        toolName: toolName,
        ruleContent: ruleContent,
        type: 'file_glob',
      };
    } else if (['webfetch', 'websearch'].includes(toolName)) {
      return {
        toolName: toolName,
        ruleContent: ruleContent.toLowerCase(),
        type: 'domain_pattern',
      };
    } else if (['glob', 'grep'].includes(toolName)) {
      return {
        toolName: toolName,
        ruleContent: ruleContent,
        type: 'search_pattern',
      };
    }

    return {
      toolName: toolName,
      ruleContent: ruleContent,
      type: 'unknown_pattern',
    };
  }

  return { toolName: toolStr.toLowerCase(), ruleContent: null, type: 'tool_name' };
}

/**
 * Match Bash command against permission pattern
 * @param {string} command - Bash command to check
 * @param {string} ruleContent - Rule content from pattern
 * @param {string} type - Pattern type (bash_wildcard or bash_exact)
 * @returns {boolean} Whether command matches pattern
 */
export function matchesBashPattern(command, ruleContent, type) {
  const cmd = command.trim().toLowerCase();
  const rule = ruleContent.toLowerCase();

  if (type === 'bash_wildcard') {
    const basePattern = rule.split(':')[0];
    const cmdParts = cmd.split(/\s+/);
    return cmdParts[0] === basePattern;
  } else {
    return cmd === rule || cmd.startsWith(rule + ' ');
  }
}

/**
 * Match file path against glob pattern
 * @param {string} filePath - File path to check
 * @param {string} globPattern - Glob pattern
 * @returns {boolean} Whether path matches pattern
 */
export function matchesFileGlob(filePath, globPattern) {
  return matchGlob(globPattern, filePath);
}

/**
 * Match URL domain against pattern
 * @param {string} url - URL to check
 * @param {string} domainPattern - Domain pattern (supports wildcards)
 * @returns {boolean} Whether domain matches pattern
 */
export function matchesDomainPattern(url, domainPattern) {
  try {
    const urlObj = new URL(url);
    const hostname = urlObj.hostname.toLowerCase();
    const pattern = domainPattern.toLowerCase();

    if (hostname === pattern) {
      return true;
    }

    if (pattern.startsWith('*.')) {
      const baseDomain = pattern.slice(2);
      return hostname === baseDomain || hostname.endsWith('.' + baseDomain);
    }

    return false;
  } catch {
    return false;
  }
}

/**
 * Check if tool matches permission string (unified for all tools)
 * @param {string} toolName - Tool name to check
 * @param {Object} input - Tool input parameters
 * @param {string} permissionStr - Permission string
 * @returns {boolean} Whether tool matches permission
 */
export function matchesPermission(toolName, input, permissionStr) {
  try {
    const parsed = parseToolPattern(permissionStr);

    if (parsed.type === 'tool_name') {
      const permToolName = parsed.toolName;
      const actualToolName = toolName.toLowerCase();

      if (permToolName.endsWith('*')) {
        const prefix = permToolName.slice(0, -1);
        return actualToolName.startsWith(prefix);
      }

      return actualToolName === permToolName;
    }

    if (toolName.toLowerCase() !== parsed.toolName) {
      return false;
    }

    switch (parsed.type) {
      case 'bash_wildcard':
      case 'bash_exact':
        return input.command
          ? matchesBashPattern(input.command, parsed.ruleContent, parsed.type)
          : false;

      case 'file_glob':
        return input.file_path ? matchesFileGlob(input.file_path, parsed.ruleContent) : false;

      case 'domain_pattern':
        return input.url ? matchesDomainPattern(input.url, parsed.ruleContent) : false;

      case 'search_pattern':
        return input.pattern ? input.pattern === parsed.ruleContent : false;

      default:
        return false;
    }
  } catch (error) {
    const errorMsg = `Permission matching failed for ${toolName} with pattern ${permissionStr}: ${error.message}`;
    console.error('[ERROR]', errorMsg, error.stack);

    console.log(
      JSON.stringify({
        type: 'error',
        message: errorMsg,
      })
    );

    return false;
  }
}

File: lua/vibing/infrastructure/permissions/matchers.lua
--- Permission pattern matching functions
--- Handles glob patterns, bash commands, domains, and tool permission strings
--- @module vibing.infrastructure.permissions.matchers

local M = {}

--- Parse URL and extract hostname
--- @param url string
--- @return string|nil hostname
local function parse_url_hostname(url)
  if type(url) ~= "string" then
    return nil
  end
  local hostname = url:match("^https?://([^/:]+)")
  if hostname then
    return hostname:lower()
  end
  return nil
end

--- Simple glob pattern matching
--- @param pattern string Glob pattern (supports * and ?)
--- @param str string String to match against
--- @return boolean
function M.match_glob(pattern, str)
  if type(pattern) ~= "string" or type(str) ~= "string" then
    return false
  end

  if #pattern > 1000 then
    return false
  end

  local ok, result = pcall(function()
    local escaped = pattern:gsub("([%.%+%^%$%(%)%[%]%%])", "%%%1")
    local regex_pattern = escaped:gsub("%*", ".-"):gsub("%?", ".")
    return str:match("^" .. regex_pattern .. "$") ~= nil
  end)

  if not ok then
    return false
  end
  return result
end

--- @class ParsedToolPattern
--- @field tool_name string
--- @field rule_content string|nil
--- @field type "bash_wildcard"|"bash_exact"|"file_glob"|"domain_pattern"|"search_pattern"|"unknown_pattern"|"tool_name"

--- Parse tool permission string like "Tool(pattern)"
--- @param tool_str string
--- @return ParsedToolPattern
function M.parse_tool_pattern(tool_str)
  local tool_name, rule_content = tool_str:match("^([a-zA-Z]+)%((.+)%)$")

  if tool_name and rule_content then
    tool_name = tool_name:lower()

    if tool_name == "bash" then
      local is_wildcard = rule_content:match("^([^:]+):%*$")
      return {
        tool_name = "bash",
        rule_content = rule_content:lower(),
        type = is_wildcard and "bash_wildcard" or "bash_exact",
      }
    elseif tool_name == "read" or tool_name == "write" or tool_name == "edit" then
      return {
        tool_name = tool_name,
        rule_content = rule_content,
        type = "file_glob",
      }
    elseif tool_name == "webfetch" or tool_name == "websearch" then
      return {
        tool_name = tool_name,
        rule_content = rule_content:lower(),
        type = "domain_pattern",
      }
    elseif tool_name == "glob" or tool_name == "grep" then
      return {
        tool_name = tool_name,
        rule_content = rule_content,
        type = "search_pattern",
      }
    end

    return {
      tool_name = tool_name,
      rule_content = rule_content,
      type = "unknown_pattern",
    }
  end

  return {
    tool_name = tool_str:lower(),
    rule_content = nil,
    type = "tool_name",
  }
end

--- Match Bash command against permission pattern
--- @param command string
--- @param rule_content string
--- @param pattern_type "bash_wildcard"|"bash_exact"
--- @return boolean
function M.matches_bash_pattern(command, rule_content, pattern_type)
  local cmd = vim.trim(command):lower()
  local rule = rule_content:lower()

  if pattern_type == "bash_wildcard" then
    local base_pattern = rule:match("^([^:]+)")
    local cmd_parts = vim.split(cmd, "%s+", { trimempty = true })
    return cmd_parts[1] == base_pattern
  else
    return cmd == rule or vim.startswith(cmd, rule .. " ")
  end
end

--- Match file path against glob pattern
--- @param file_path string
--- @param glob_pattern string
--- @return boolean
function M.matches_file_glob(file_path, glob_pattern)
  return M.match_glob(glob_pattern, file_path)
end

--- Match URL domain against pattern
--- @param url string
--- @param domain_pattern string
--- @return boolean
function M.matches_domain_pattern(url, domain_pattern)
  local hostname = parse_url_hostname(url)
  if not hostname then
    return false
  end

  local pattern = domain_pattern:lower()

  if hostname == pattern then
    return true
  end

  if vim.startswith(pattern, "*.") then
    local base_domain = pattern:sub(3)
    return hostname == base_domain or vim.endswith(hostname, "." .. base_domain)
  end

  return false
end

--- Check if tool matches permission string (unified for all tools)
--- @param tool_name string
--- @param input table<string, any>
--- @param permission_str string
--- @return boolean
function M.matches_permission(tool_name, input, permission_str)
  local ok, result = pcall(function()
    local normalized_permission_str = permission_str:gsub(":once$", "")
    local parsed = M.parse_tool_pattern(normalized_permission_str)

    if parsed.type == "tool_name" then
      local perm_tool_name = parsed.tool_name
      local actual_tool_name = tool_name:lower()

      if vim.endswith(perm_tool_name, "*") then
        local prefix = perm_tool_name:sub(1, -2)
        return vim.startswith(actual_tool_name, prefix)
      end

      return actual_tool_name == perm_tool_name
    end

    if tool_name:lower() ~= parsed.tool_name then
      return false
    end

    if parsed.type == "bash_wildcard" or parsed.type == "bash_exact" then
      if input.command then
        return M.matches_bash_pattern(input.command, parsed.rule_content, parsed.type)
      end
      return false
    elseif parsed.type == "file_glob" then
      if input.file_path then
        return M.matches_file_glob(input.file_path, parsed.rule_content)
      end
      return false
    elseif parsed.type == "domain_pattern" then
      if input.url then
        return M.matches_domain_pattern(input.url, parsed.rule_content)
      end
      return false
    elseif parsed.type == "search_pattern" then
      if input.pattern then
        return input.pattern == parsed.rule_content
      end
      return false
    end

    return false
  end)

  if not ok then
    vim.notify(
      string.format("Permission matching failed for %s with pattern %s: %s", tool_name, permission_str, result),
      vim.log.levels.ERROR
    )
    return false
  end

  return result
end

return M
